Incorporate isomorphism testing for edge coloured graphs

markusbaumeister opened the pull request:
There are two different types of nauty-graphs. Those without edge colouring store their edges in the component "edges", but those with edge colouring store them in the component "edge_list". Up to this point, the methods "CanonicalForm" and "IsomorphismGraphs" were only implemented for graphs without edge colouring.

markusbaumeister commented:
The travis build fails in 4.11 and master because of a change in Digraphs. More concretely, the error is
``
Petersen := Digraph( Petersen );
# Expected output:
<digraph with 10 vertices, 30 edges>
# But found:
<immutable digraph with 10 vertices, 30 edges>
``

fingolfin commented:
@aniemeyer just asked me to help out with this.

This PR needs to be rebased; would you be willing to do that, @markusbaumeister ? Alternatively I can try to do that, but I don't want to do this "behind your back".

markusbaumeister commented:
@fingolfin Rebase complete. 

FYI, I would not have minded if you went ahead here, since you wouldn't have broken any of my work ;)

fingolfin commented on examples/test.g:
```diff
@@ -3,7 +3,7 @@ LoadPackage( "Grape" );
 Petersen := Graph( SymmetricGroup(5), [[1,2]], OnSets,
                    function(x,y) return Intersection(x,y)=[]; end );
-Petersen := Digraph( Petersen );
+Petersen := Digraph( Petersen );;
```
What's the motivation for this?

markusbaumeister replied:
I don't recall. Maybe I tried to fix problems that arose due to the depiction of the Digraph-package? I reverted that change.

fingolfin commented on gap/NautyGraph.gi:
```diff
@@ -361,6 +361,40 @@ InstallMethod( CanonicalForm,
+    perm := CanonicalLabeling( graph );
+    permInv := CanonicalLabelingInverse( graph );
+    
+    edges := graph!.edge_list;
+
+    permEdges := [];
+    # Since this computation should be efficient we do it manually
+    for edgeClass in edges do
+        newClass := [];
+
+        for i in [ 1 .. Length( edgeClass ) ] do # edges is a dense list
+            # Every line of code in this loop costs a lot of time
+            Add( newClass, [ edgeClass[ i ][ 1 ]^permInv, edgeClass[ i ][ 2 ]^permInv ] );
+            if newClass[ i ][ 1 ] > newClass[ i ][ 2 ] then
+                newClass[ i ] := [ newClass[ i ][ 2 ], newClass[ i ][ 1 ] ];
+            fi;
+        od;
+
+        Add(permEdges, Set(newClass) );
+    od;
+    edges := permEdges;
```
Or you could do this instead:
```suggestion
    edges := List( graph!.edge_list, edgeClass -> OnSetsSets( edgeClass, permInv ) );
```

markusbaumeister replied:
Yes, but that would be slower. Since this loop is called many times, code that computes many isomorphisms is significantly speed up by the manual computation.

fingolfin replied:
Have you timed it? In any case, if you are concerned about performance, you should still consider replacing the body of the innermost loop by using `OnSets`, which is a kernel function and hard to beat with plain GAP code.

fingolfin reviewed (commented):
Seems plausible enough, but it would be really nice to have some tests, even if they are trivial; the new code is not called at all right now, it seems.

markusbaumeister commented:
Unfortunately, I do not know how to efficiently generate tst-files. Since all tst-files so far are mirrored in the examples-directory, I tried adding the test there. Apparently that did not do the trick. How to I generate tst-files automatically?

fingolfin commented on gap/NautyGraph.gi:
```diff
@@ -361,6 +361,40 @@ InstallMethod( CanonicalForm,
+    for edgeClass in edges do
+        newClass := [];
+
+        for i in [ 1 .. Length( edgeClass ) ] do # edges is a dense list
+            # Every line of code in this loop costs a lot of time
+            Add( newClass, [ edgeClass[ i ][ 1 ]^permInv, edgeClass[ i ][ 2 ]^permInv ] );
+            if newClass[ i ][ 1 ] > newClass[ i ][ 2 ] then
+                newClass[ i ] := [ newClass[ i ][ 2 ], newClass[ i ][ 1 ] ];
+            fi;
+        od;
```
As mentioned, I would recommend also benchmarking this variant, as `OnSets` is a kernel function (which of course handles arbitrary sets, not just sets of size 2, so it is not as efficient as possible, but it still might beat the current code)
```suggestion
        for edge in edgeClass do # edges is a dense list
            # Every line of code in this loop costs a lot of time
            Add( newClass, OnSets( edge, permInv ) );
        od;
```
I would also try this:
```suggestion
        newClass := List( edgeClass, edge -> OnSets( edge, permInv ) );
```

fingolfin replied:
Even if `OnSets` turns out not to be faster, you definitely can improve the code by using local variables: i.e. something along these lines:
```gap
a:= edge[1]^permInv;
b:= edge[2]^permInv;
if a < b then
   Add(newClass, [a,b]);
else
   Add(newClass, [b,a]);
fi;
```

fingolfin commented on examples/edgeColoured.g:
```diff
@@ -0,0 +1,6 @@
+
+IsomorphismGraphs(g1,g2);
+
+Assert( IsomorphismGraphs(g1,g2) = (1,2)(3,5)(4,9,10,12,8)(6,11), "Edge-coloured isomorhism fails");
```
To turn this into a test file, basically prefix all input lines with `gap> `. Then either already fill in the expected return values; resp. suppress them using `;;` if they are uninteresting and long. Or let GAP do the work by using `Test("FILENAME.tst", rec(rewriteToFile:=true));`. 

That way I came up with this:
```
gap> g1 := NautyEdgeColoredGraph( [[[1,2],[3,10],[11,12],[4,9],[5,6],[7,8]],[[1,12],[2,3],[10,11],[4,5],[6,7],[8,9]],[[3,4],[9,10]]],12);;
gap> g2 := NautyEdgeColoredGraph( [[[3,11],[9,10],[4,7],[1,2],[5,12],[6,8]],[[7,11],[4,10],[3,9],[1,5],[2,8],[6,12]],[[5,9],[10,12]]],12);;
gap> IsomorphismGraphs(g1,g2);
(1,2)(3,5)(4,9,10,12,8)(6,11)
```

markusbaumeister commented:
Concerning benchmarking: Is there a standard way to record such benchmarks for different code variants?

fingolfin commented on gap/NautyGraph.gi:
```diff
@@ -361,6 +361,40 @@ InstallMethod( CanonicalForm,
+    local colors, edges, perm, new_graph, permEdges, i,
+        permInv, edgeClass, newClass;
+    
+    perm := CanonicalLabeling( graph );
```
BTW is `perm` even used here?

markusbaumeister replied:
Good catch. The same is true for the other Canonical-method. I removed both.

fingolfin commented:
I am not sure what you are looking for, specifically? Benchmarking overall is a complex art. For GAP code, I sometimes use this code for benchmarking individual code: <https://gist.github.com/fingolfin/b81101cc304d6062d195fa80fae32ab4>. I also try to measure on a quiet and idle server (and not e.g. my laptop, which produces massive performance fluctuations). Also https://github.com/sharkdp/hyperfine is very useful.

Luckily the relevant code is easy to extract and test in isolation. But what would be good test data? I've tried your example, but this is so tiny that all the various variants of the code essentially take the same amount of time (close to zero). Here's my test:
```gap
f1:=function(edges, permInv)
    local permEdges, edgeClass, newClass, i;
    permEdges := [];
    for edgeClass in edges do
        newClass := [];

        for i in [ 1 .. Length( edgeClass ) ] do # edges is a dense list
            Add( newClass, [ edgeClass[ i ][ 1 ]^permInv, edgeClass[ i ][ 2 ]^permInv ] );
            if newClass[ i ][ 1 ] > newClass[ i ][ 2 ] then
                newClass[ i ] := [ newClass[ i ][ 2 ], newClass[ i ][ 1 ] ];
            fi;
        od;

        Add(permEdges, Set(newClass) );
    od;
    return permEdges;
end;

f2:=function(edges, permInv)
    local permEdges, edgeClass, newClass, edge, a, b;
    permEdges := [];
    for edgeClass in edges do
        newClass := [];

        for edge in edgeClass do # edges is a dense list
            a := edge[1]^permInv;
            b := edge[2]^permInv;
            if a < b then
                Add(newClass, [a,b]);
            else
                Add(newClass, [b,a]);
            fi;
        od;

        Add(permEdges, Set(newClass) );
    od;
    return permEdges;
end;

f3:=function(edges, permInv)
    local permEdges, edgeClass, newClass, edge;
    permEdges := [];
    for edgeClass in edges do
        newClass := [];

        for edge in edgeClass do # edges is a dense list
            Add(newClass, OnSets(edge, permInv));
        od;

        Add(permEdges, Set(newClass) );
    od;
    return permEdges;
end;

f4:=function(edges, permInv)
    local permEdges, edgeClass;
    permEdges := [];
    for edgeClass in edges do
        Add(permEdges, Set(edgeClass, edge -> OnSets(edge, permInv)));
    od;
    return permEdges;
end;

f5:=function(edges, permInv)
    return List(edges, edgeClass -> Set(edgeClass, edge -> OnSets(edge, permInv)));
end;

f6:=function(edges, permInv)
    return List( edges, edgeClass -> OnSetsSets( edgeClass, permInv ) );;
end;


LoadPackage("nauty");


graph := NautyEdgeColoredGraph( [[[1,2],[3,10],[11,12],[4,9],[5,6],[7,8]],[[1,12],[2,3],[10,11],[4,5],[6,7],[8,9]],[[3,4],[9,10]]],12);;
edges := graph!.edge_list;
permInv := CanonicalLabelingInverse( graph );

Assert(0, 1 = Length(Set([f1,f2,f3,f4,f5,f6], f -> f(edges, permInv))));

Benchmark(function() f1(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Benchmark(function() f2(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Benchmark(function() f3(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Benchmark(function() f4(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Benchmark(function() f5(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Benchmark(function() f6(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
```
and this is what I get:
```
gap> Benchmark(function() f1(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Performed 100 iterations, taking 3. milliseconds.
average: 0.03 +/- 0. (+/- 9.%)
median: 0.02837599813938141
gap> Benchmark(function() f2(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Performed 100 iterations, taking 3. milliseconds.
average: 0.03 +/- 0. (+/- 16.%)
median: 0.02563949674367905
gap> Benchmark(function() f3(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Performed 100 iterations, taking 3. milliseconds.
average: 0.03 +/- 0. (+/- 10.%)
median: 0.02620051056146622
gap> Benchmark(function() f4(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Performed 100 iterations, taking 3. milliseconds.
average: 0.03 +/- 0. (+/- 12.%)
median: 0.03133400529623032
gap> Benchmark(function() f5(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Performed 100 iterations, taking 4. milliseconds.
average: 0.04 +/- 0. (+/- 12.%)
median: 0.03491249680519104
gap> Benchmark(function() f6(edges, permInv); end, rec(maxtime:=1000, maxreps:=100));;
Performed 100 iterations, taking 3. milliseconds.
average: 0.03 +/- 0. (+/- 9.%)
median: 0.03277800232172012
```

fingolfin commented on gap/NautyGraph.gi:
```diff
@@ -373,23 +407,37 @@ BindGlobal( "NAUTYTRACESINTERFACE_IsomorphismGraphsOnlyEdges",
+        return CanonicalLabelingInverse( graph1 ) * CanonicalLabeling( graph2 );
+    fi;
+    
+    return fail;
```
The coverage failure here indicates that a test is missing which tests two non-isomorphic graphs for being isomorphic, I think?

fingolfin commented on tst/edgeColoured.tst:
```diff
@@ -0,0 +1,4 @@
+gap> g1 := NautyEdgeColoredGraph( [[[1,2],[3,10],[11,12],[4,9],[5,6],[7,8]],[[1,12],[2,3],[10,11],[4,5],[6,7],[8,9]],[[3,4],[9,10]]],12);;
+gap> g2 := NautyEdgeColoredGraph( [[[3,11],[9,10],[4,7],[1,2],[5,12],[6,8]],[[7,11],[4,10],[3,9],[1,5],[2,8],[6,12]],[[5,9],[10,12]]],12);;
+gap> IsomorphismGraphs(g1,g2);
+(1,2)(3,5)(4,9,10,12,8)(6,11)
```
So we should also have a test here for two non-isomorphic graphs. And perhaps some corner/edge cases, like what happens if the two graphs have different numbers of vertices.

fingolfin replied:
(I can also help out with that, just don't want to forget it)

markusbaumeister commented:
I added some of the tests you asked for.

I also added a benchmark-directory, to make it easier to compare the benchmarks against each other. On my Laptop (unfortunately, I do not have something better to run these tests on), the method f3 performed best. You can find my results in benchmarks/Benchmark_...

To generate the necessary examples, I used the twisted-branch of the SimplicialSurfaces-package (https://github.com/markusbaumeister/simplicial-surfaces/pull/119). Since that one is not easily usable before this PR is resolved, I also generated the examples as an independently usable list.

I hope that resolves all of your comments so far (they are a bit scattered ;) ).

fingolfin commented on .gitignore:
```diff
@@ -26,3 +26,4 @@
 /configure
 /gen/
 /Makefile
+*swp
```
This conflicts with master.

I would recommend to add this to your personal global gitignore instead (see e.g. https://stackoverflow.com/a/5724484/928031 for how to create one), as not everybody is using your editor, while you are probably using it everywhere; so it's better to exclude it for every repo you work with?

fingolfin commented:
So remarkably, using `OnSetsSets`, while not being the fastest, still beats your original "optimized" code -- that's not meant as a slight, BTW, just to emphasis what I said earlier, namely that using the profiler to find hotspots is good, but then tuning the hotspots shouldn't involve the profiler (but of course after the tuning has been done one can run the profiler again to see if there are new hotspots)

markusbaumeister commented:
Unfortunately, that has ramifications for the remainder of the code base. Some of that was also fine-tuned with the profiler :/. But changing that code is probably better done in a separate PR.

fingolfin commented:
I don't quite feel comfortable with adding this particular "benchmark": it's a throw-away test of one isolated & special loop, it doesn't seem worthwhile keeping that around. So I've rebased this PR, squashed it and dropped the benchmarks; will merge the result now.

I've also keep a backup copy of the old form the of branch resp. the benchmark files, though, i.e., no data was lost :-).

Thank you very much for this PR!
